### ISSUE TITLE
Nudge users who are not entering data

### ISSUE THREAD

**Issue by mahalakshme** [avniproject/avni-server#825]
### Requirement:
Nudge users who are have not entered data for past 5 days

### AC:
- Do in integration service to retain sanity and not to corrupt ETL
- Store custom_query to retrieve users who have not entered data for past 5 days. 
- Call the glific API(on Avni) synchrously from integration-service and store the error records in integration service. One user at a time message can be sent.
- Have a window such that the job should not send message to the same users for whom message was sent in the past 23 hours. So somewhere need to store the users data for whom message was sent in the past 23 hours.
- Store the manual_message in table
- Currently we can retain the 4 days timeline to retry for any failed messages.
- For any other missing details refer [this](https://github.com/avniproject/avni-server/issues/836) spike. Also checkout the commits to implement or cherry-pick.

### AC:
- Have a ETL/integration service job to identify the users who meet all the following criteria:
        - Didn't enter/edit any registrations or encounters in the past 5 days
        - The number of days should be configurable easily
        - Should belong to 'Primary Users' user group
        - Should not consider users who have completed endline for all the work orders in their catchment
- Store the users identified above in some ETL/some table temporarily(Needed just for testability)

### Technical suggestions:
- Option 1: integration service that executes custom query to figure out the users.
- Option 2: ETL bg job that uses ETL tables to figure out the users
- Option 3: ETL bg job that uses custom query to figure out the users

Approaches discarded because of complexity or suitability: Background job in Avni, Using Avni APIs, Dalgo

Independent of the approach followed, the same ETL and integration service should be usable in Avni as well without forking.

### Out of scope:
Sending messages to such users using Glific.

### Inputs:
- int., service access to ETL database?
- window to not send msg again - 1 day - run every 10 mins 
- - report - to whom sent/failed message
- - job monitoring
- /executeQuery?
- add spike 
- add 30 percent estimate

**Comment by himeshr** [comment]
## Possible issue with AC

### Condition set 1
- Call the glific API(on Avni) synchrously from integration-service and store the error records in integration service. One user at a time message can be sent.
- Have a window such that the job should not send message to the same users for whom message was sent in the past 23 hours. So somewhere need to store the users data for whom message was sent in the past 23 hours.

### Condition set 2
- Store the manual_message in table
- Currently we can retain the 4 days timeline to retry for any failed messages.


If we do conditionSet 1, there is no need to do conditionSet 2, as we'll run the nudge main job to check and send message once every day in a synchronous manner.

**Comment by himeshr** [comment]
Putting this card on Hold, as we are facing following 2 issues:

- **ATECF Glific Application Access**
The Glific application is currently inaccessible, and we’re encountering a "403 Forbidden" error when trying to log in. This impedes our development and testing. User credentials used are for +919632322066 login.

- **Parameters for Nudge Users Glific Message Template**
We would also like to know which parameters need to be passed to the Nudge Users Glific Message Template. For example, in the case of the "User Registration" Glific Message Template (Template-id: 333223), we pass the UserId and Password as parameters.

**Comment by himeshr** [comment]
## Deployment steps

- Insert customQuery for targeted RWB org..
```sql
insert into custom_query (uuid, name, query, organisation_id, is_voided, version, created_by_id,
                          last_modified_by_id, created_date_time, last_modified_date_time)

values (uuid_generate_v4(),
       'Inactive users', 'WITH impacted_users as (
    WITH open_work_orders as (select i.id open_wo_id, i.address_id
                              from individual i
                                       left join encounter e on i.id = e.individual_id
                              where i.is_voided = false
                                and i.subject_type_id = (select id from subject_type where name = ''Work Order'' and organisation_id = (select id from organisation where db_user = ''rwbniti'') and not subject_type.is_voided)
                                and e.encounter_type_id =
                                    (select id from encounter_type where name = ''Work order endline''  and organisation_id = (select id from organisation where db_user = ''rwbniti'') and not encounter_type.is_voided)
                                and (e.is_voided is null or e.is_voided = false)
                                and i.organisation_id = (select id from organisation where db_user = ''rwbniti'')
                              group by 1
                              having count(e.id) = 1) -- Use = 1 for testing, < 1 for prod
    select distinct u.id user_id, u.username first_name
    from open_work_orders owo
             join virtual_catchment_address_mapping_table cam on cam.addresslevel_id = owo.address_id
             join users u on u.catchment_id = cam.catchment_id and u.is_voided = false
             join user_group ug on u.id = ug.user_id and ug.is_voided = false
             join groups g on g.id = ug.group_id and g.is_voided = false
    where g.name = ''Primary Users''
),
     active_user_ids as (select (case
                                     when ind.created_date_time > TO_TIMESTAMP(:cutOffDate, ''YYYY-MM-DDTHH24:MI:ss.MS'')
                                         then ind.created_by_id end) as cuid,
                                ind.last_modified_by_id              as muid
                         from individual ind
                         where ind.last_modified_date_time > TO_TIMESTAMP(:cutOffDate, ''YYYY-MM-DDTHH24:MI:ss.MS'')
                         UNION
                         select (case
                                     when enc.created_date_time > TO_TIMESTAMP(:cutOffDate, ''YYYY-MM-DDTHH24:MI:ss.MS'')
                                         then enc.created_by_id end) as cuid,
                                enc.last_modified_by_id              as muid
                         from encounter enc
                         where enc.last_modified_date_time > TO_TIMESTAMP(:cutOffDate, ''YYYY-MM-DDTHH24:MI:ss.MS'')
                         order by 1 asc)
select distinct user_id, first_name
from impacted_users iu
where user_id not in (select cuid
                      from active_user_ids
                      union
                      select muid
                      from active_user_ids);',
        21,
        false,
        0,
        1,
        1,
        now(),
        now());
```

**Comment by mahalakshme** [comment]
@himeshr this is unblocked. Credentials in Keeweb. Also tested rwb2023 org in prerelease with the configuration. Template to be used is 'User Nudge'

<img width="885" alt="Image" src="https://github.com/user-attachments/assets/ab5bbf02-2ae8-45b7-8d17-0b7613a5e5e0" />

Multiple 'User Nudge' templates will show up. Selection of any should work. Each one created for different languages. All templates in different languages are mapped to same id. Hence based on the language configured for the user, message will get sent automatically in that language - feature of Glific - tested and it is working.

**Comment by mahalakshme** [comment]
@himeshr have added the parameter details in the above comment. Missed it.

**Comment by petmongrels** [comment]
- We won't be able to run this query in QA environments as the org name is hardcoded. One way we can solve this and the security issue in general is to make :organisation_id  or :db_user as a mandatory parameter in custom queries. The platform can always set the organisation_id/dbuser in the prepared statement. All custom queries must have one of this parameter.
- cc @1t5j0y

**Comment by himeshr** [comment]
> * We won't be able to run this query in QA environments as the org name is hardcoded. One way we can solve this and the security issue in general is to make :organisation_id  or :db_user as a mandatory parameter in custom queries. The platform can always set the organisation_id/dbuser in the prepared statement. All custom queries must have one of this parameter.
> * cc [@1t5j0y](https://github.com/1t5j0y)

Will include both organisation id, schema_name and db_user as additional queryParams.

**Comment by himeshr** [comment]
## Query to execute to fetch list of users that are to be nudged

##### QueryParams
- org_db_user => 'rwbni**'
- org_id => 191
- cutOffDate => '2025-02-07'

```
WITH impacted_users as (
    WITH open_work_orders as (select i.id open_wo_id, i.address_id
                              from individual i
                                       left join encounter e on i.id = e.individual_id
                              where i.is_voided = false
                                and i.subject_type_id = (select id from subject_type where name = 'Work Order' and organisation_id = (select id from organisation where db_user = :org_db_user) and not subject_type.is_voided)
                                and e.encounter_type_id =
                                    (select id from encounter_type where name = 'Work order endline'  and organisation_id = (select id from organisation where db_user = :org_db_user) and not encounter_type.is_voided)
                                and (e.is_voided is null or e.is_voided = false)
                                and i.organisation_id = :org_id
                              group by 1
                              having count(e.id) = 1) -- Use = 1 for testing, < 1 for prod
    select distinct u.id user_id, u.name first_name
    from open_work_orders owo
             join virtual_catchment_address_mapping_table cam on cam.addresslevel_id = owo.address_id
             join users u on u.catchment_id = cam.catchment_id and u.is_voided = false
             join user_group ug on u.id = ug.user_id and ug.is_voided = false
             join groups g on g.id = ug.group_id and g.is_voided = false
    where g.name = 'Primary Users'
),
     active_user_ids as (select (case
                                     when ind.created_date_time > TO_TIMESTAMP(:cutOffDate, 'YYYY-MM-DDTHH24:MI:ss.MS')
                                         then ind.created_by_id end) as cuid,
                                ind.last_modified_by_id              as muid
                         from individual ind
                         where ind.last_modified_date_time > TO_TIMESTAMP(:cutOffDate, 'YYYY-MM-DDTHH24:MI:ss.MS')
                         UNION
                         select (case
                                     when enc.created_date_time > TO_TIMESTAMP(:cutOffDate, 'YYYY-MM-DDTHH24:MI:ss.MS')
                                         then enc.created_by_id end) as cuid,
                                enc.last_modified_by_id              as muid
                         from encounter enc
                         where enc.last_modified_date_time > TO_TIMESTAMP(:cutOffDate, 'YYYY-MM-DDTHH24:MI:ss.MS')
                         order by 1 asc)
select distinct user_id, first_name
from impacted_users iu
where user_id not in (select cuid
                      from active_user_ids
                      union
                      select muid
                      from active_user_ids);
```

**Comment by mahalakshme** [comment]
@himeshr We need not insert message_rule and all? We need to mention message template via some configuration?

**Comment by mahalakshme** [comment]
Himesh reply: No insertion of messageRule, we just have to specify the messageTemplateID in integration rwb-app properties..

rwb.avni.message.template.id=${RWB_AVNI_MESSAGE_TEMPLATE_ID:542201}

**Comment by mahalakshme** [comment]
@himeshr I didnt yet receive any nudge message, though my user exists in 'RWB 2024 Staging' organisation and I have not done any work in the past one month.

**Comment by 1t5j0y** [comment]
FYI @himeshr and @mahalakshme I just received what looks like a 'nudge' message.

![Image](https://github.com/user-attachments/assets/ac8ddf21-59bc-46b2-9985-0f0152fab01c)

**Comment by mahalakshme** [comment]
@1t5j0y Yayy!! Saw in the glific webapp.  I also received around the same time.

**Comment by mahalakshme** [comment]
Cases tested with one day window:
- Send msg to all users in the 'Primary User group' when no work done in the past many days
- Farmer registered in the past one day - then dont send message
- Work order registered and endline completed 5 days back - then dont send message
- Send message again after a day when no work done for past few days

**Comment by himeshr** [comment]
@mahalakshme Though Joy's profile has English language, message is sent in Hindi
It was supposed to internally use user profile specific template and send the msg right?

> FYI [@himeshr](https://github.com/himeshr) and [@mahalakshme](https://github.com/mahalakshme) I just received what looks like a 'nudge' message.
> 
> ![Image](https://github.com/user-attachments/assets/ac8ddf21-59bc-46b2-9985-0f0152fab01c)

![Image](https://github.com/user-attachments/assets/02ac0c8e-f857-4474-907d-48cdb40e4600)

**Comment by mahalakshme** [comment]
@himeshr Yes have sent a mail to Glific team asking about this. Found the issue in Maha contact when I was testing as well.

**Comment by himeshr** [comment]
- Fixed the query with [this ](https://github.com/avniproject/integration-service/commit/8d5e68c80293e6549eab59ad9f924dadd46b3319) commit. The changed sql query needs to be applied on the gdgs25-26 uat org.

- We need to restart integration-service post config change for it to be reflected.

**Comment by AchalaBelokar** [comment]
- [ ] I register work order on 25 feb after that nothing will happen with acha@gdgs25-26uat  I didn’t get any message

**Comment by himeshr** [comment]
Has few configuration mismatch due to recent changes to the apiuser:
- Correction in password on integration-service
- Correction in Glific Users usergroup to "Enable" "Messaging" privilege, to be able to invoke the syncSendMsg call

**Comment by AchalaBelokar** [comment]
- [ ] I created a user name achalab@gdgs25-26uat and then created work order with date 3.3.2025 but still I didn’t receive any message

**Comment by himeshr** [comment]
Made changes to query to ignore null values, which mess up in clause check.

**Comment by AchalaBelokar** [comment]
- [ ] I created one user name is achala123@gdgs25-26uat earlier it was not a primary user and also I change the mobile no. and then I created  work order and last modification is 8 march still I didn’t get any message.

**Comment by himeshr** [comment]
> * [ ]  I created one user name is achala123@gdgs25-26uat earlier it was not a primary user and also I change the mobile no. and then I created  work order and last modification is 8 march still I didn’t get any message.

For this user, there is individual shown as recently modified..
id,id,uuid
48243,48243,f82c4298-5ce4-4afc-849a-6efa6ad04142
